### PR TITLE
Typecast bool to int

### PR DESCRIPTION
**Pull request by SignpostMarv**
re: #2548, not sure which test case to add tests to (if needed) ?

### CODE REVIEW

**Comment by SignpostMarv** [comment]
> `node_data->setType($stmt->expr);`

is this not meant to be `getType($stmt->expr)` ?

**Comment by muglug** [comment]
oops yes, corrected

**Comment by SignpostMarv** [comment]
Does this look about right?

```diff
diff --git a/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php b/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php
index 54b9b989..1382c6ac 100644
--- a/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php
@@ -553,34 +553,15 @@ class ExpressionAnalyzer
         } elseif ($stmt instanceof PhpParser\Node\Expr\Cast\Int_) {
             if (self::analyze($statements_analyzer, $stmt->expr, $context) === false) {
                 return false;
-            } elseif (($stmt->expr instanceof PhpParser\Node\Expr\ConstFetch) &&
-                (
-                    ['true'] === $stmt->expr->name->parts ||
-                    ['false'] === $stmt->expr->name->parts
-                )
-            ) {
-                $int = 0;
-
-                if (['true'] === $stmt->expr->name->parts) {
-                    $int = 1;
-                }
-
-                $statements_analyzer->node_data->setType($stmt, Type::getInt(false, $int));
-            } elseif ($stmt->expr instanceof PhpParser\Node\Expr\Cast\Bool_) {
-                if ($stmt->expr->expr instanceof PhpParser\Node\Scalar\LNumber) {
-                    $int = 0;
+            }

-                    if (0 !== $stmt->expr->expr->value) {
-                        $int = 1;
-                    }
+            $maybe_bool = $statements_analyzer->node_data->getType($stmt->expr);

-                    $statements_analyzer->node_data->setType($stmt, Type::getInt(false, $int));
-                } else {
+            if (1 === count($maybe) && $maybe_bool[0] instanceof Type\Atomic\TBool) {
                     $statements_analyzer->node_data->setType($stmt, new Type\Union([
                         new Type\Atomic\TLiteralInt(0),
                         new Type\Atomic\TLiteralInt(1),
                     ]));
-                }
             } else {
                 $statements_analyzer->node_data->setType($stmt, Type::getInt());
             }
```

**Comment by muglug** [comment]
`NodeTypeProvider::getType` returns a union type. `Union::getTypes` returns atomic types, so you need to call both

**Comment by muglug** [comment]
I've just renamed `Union::getTypes` to `Union::getAtomicTypes` to make things a bit more explicit: 9fa2db1

**Comment by muglug** [comment]
Also write tests! I think these could go in `ValueTest` or `AssignmentTest`

**Comment by SignpostMarv** [comment]
> you need to call both

i.e. ? 

```diff
+            $maybe = $statements_analyzer->node_data->getType(
+                $stmt->expr
+            )->getAtomicTypes();
```

**Comment by muglug** [comment]
yes, but also (in rare circumstances) `$maybe` can return `null`, so you need to check for that.

**Comment by SignpostMarv** [comment]
> so you need to check for that.

forgetting how you feel about inline assignments:
```diff
+            if (
+                ! is_null($maybe_type) &&
+                1 === count($maybe = $maybe_type->getAtomicTypes()) &&
+                current($maybe) instanceof Type\Atomic\TBool
+            ) {
```

**Comment by muglug** [comment]
I prefer explicit `!== null` checks and yes, inline assignments are generally to be avoided – especially inside function calls. Easily-scannable code is more maintainable code.

**Comment by SignpostMarv** [comment]
yeah just ignore 698f94e, I had the wrong file open ¬_¬

**Comment by SignpostMarv** [comment]
is a test that catches stuff like `random_int(0, 2)` necessary ?

File: tests/TypeReconciliation/ValueTest.php [hunk 1]
```diff
@@ -656,6 +656,39 @@ class A {
...
+                /**
+                * @param 0|1 $int
+                */
+                function foo(int $int) : void {
+                    echo (string) $int;
+                }
+
+                foo((int) true);',
+            ],
+            'typecastFalseToInt' => [
```
Comment (muglug) [inline tests/TypeReconciliation/ValueTest.php]: I'd say this test isn't necessary

**Comment by muglug** [comment]
> is a test that catches stuff like random_int(0, 2) necessary ?

No

**Comment by muglug** [comment]
Thanks!

**Pull request merged**